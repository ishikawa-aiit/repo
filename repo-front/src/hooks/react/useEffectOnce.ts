import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
