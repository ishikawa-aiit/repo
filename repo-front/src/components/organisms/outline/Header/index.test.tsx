import { screen, within } from '@testing-library/react'
import { useRecoilState } from 'recoil'

import { renderWithQueryClient } from '../../../../../.jest'

import { Header } from '.'

jest.mock('recoil')
const user = {
  name: '新渡戸 稲造',
}
;(useRecoilState as jest.Mock).mockReturnValue([user])

describe('snapshots', () => {
  it('最新のスナップショットに更新されている', () => {
    const { container } = renderWithQueryClient(<Header />)

    expect(container).toMatchSnapshot()
  })

  it('ログイン中はユーザー名が表示される', () => {
    renderWithQueryClient(<Header />)

    const { getByText } = within(screen.getByTestId('HeaderUserName'))
    expect(getByText(user.name)).toBeInTheDocument()
  })
})
