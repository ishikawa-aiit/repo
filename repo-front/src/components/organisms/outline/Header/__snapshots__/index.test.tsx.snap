// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots 最新のスナップショットに更新されている 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  position: -webkit-sticky;
  position: sticky;
  top: 0px;
  left: 0px;
  background: white;
  width: 100%;
  -webkit-padding-start: 5px;
  padding-inline-start: 5px;
  -webkit-padding-end: 5px;
  padding-inline-end: 5px;
  box-shadow: sm;
  z-index: 10;
}

.emotion-2 {
  padding-top: 2px;
  padding-bottom: 1px;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-2:hover,
.emotion-2[data-hover] {
  opacity: .75;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-5 {
  font-size: 80%;
  padding-left: .8rem;
}

<div>
  <header
    class="emotion-0"
    data-cy="Header"
    data-testid="Header"
  >
    <a
      class="chakra-link emotion-1"
      href="/weekly-reports/create"
    >
      <div
        class="emotion-2"
      >
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
        >
          <span
            style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27110%27%20height=%2750%27/%3e"
              style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
            />
          </span>
          <img
            alt="週報作成支援ツール Repo"
            data-nimg="intrinsic"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </div>
    </a>
    <div
      class="emotion-3"
    >
      <p
        class="chakra-text emotion-1"
        data-testid="HeaderUserName"
      >
        新渡戸 稲造
        <small
          class="chakra-text emotion-5"
        >
          さん
        </small>
      </p>
    </div>
  </header>
</div>
`;
